MaxText/decode.py

```diff
@@ -82,79 +86,99 @@ def encode_strings(strs, max_len, tokenizer, mesh):
       replicate_globally(positions, mesh), \
       replicate_globally(segment_ids, mesh)
 
+
 def prefill_predict_step(inputs, input_positions, decoder_segment_ids,
-                 state,
-                 rngkey,
-                 model=None):
+                         model_vars, rngkey, model=None):
   """Prefill KV Cache and output logits"""
+  convert_mode = quantizations.in_convert_mode(model.quant)
+  mutable = True if convert_mode else ['cache']
   flat_logits, new_vars = model.apply(
-    {
-        "params": state.params
-    },
+    model_vars,
     inputs,
     input_positions,
     decoder_segment_ids=decoder_segment_ids,
     enable_dropout=False,
     model_mode=common_types.MODEL_MODE_PREFILL,
     rngs={'params': rngkey},
-    mutable=["cache"]
+    mutable=mutable
   )
+  prefill_cache = new_vars['cache']
+  new_vars.pop('cache')
+  if not convert_mode:
+    new_vars = None
+  return flat_logits, prefill_cache, new_vars
 
-  return flat_logits, new_vars['cache']
 
-def ar_predict_single_token(previous_logits, token_position, kv_cache, state, rngkey, model, config):
+def ar_predict_single_token(previous_logits, token_position, kv_cache, model_vars, rngkey, model, config):
   """Predict one token, return new cache"""
 
   new_token = inference_utils.sampling(previous_logits, rngkey, config.decode_sampling_strategy,\
                                        topk=config.decode_sampling_top_k, nucleus_topp=config.decode_sampling_nucleus_p,
                                        temperature=config.decode_sampling_temperature)
+
+  convert_mode = quantizations.in_convert_mode(model.quant)
+  mutable = True if convert_mode else ['cache']
   flat_logits, new_vars = model.apply(
-    {
-        "params": state.params,
-        "cache": kv_cache
-    },
+    model_vars | {'cache': kv_cache},
     new_token,
     token_position,
     enable_dropout=False,
     model_mode=common_types.MODEL_MODE_AUTOREGRESSIVE,
     rngs={'params': rngkey},
-    mutable=["cache"])
+    mutable=mutable)
   new_flat_cache = new_vars["cache"]
-  return token_position+1, new_flat_cache, flat_logits, new_token
+  new_vars.pop('cache')
```

Comment:
can we convert in init? Super weird to have it spread out.

Reply:
Discussed offline - we use the prefill apply method to call a "dummy" prefill to initiate aqt state. Cleaned up the function a bit.